### PR TITLE
CTP 38 implement friction

### PR DESCRIPTION
**Pull request by squiree98**

### CODE REVIEW

**Review by Nijelous — CHANGES_REQUESTED** [review]
Overall looks good, but to keep up coding standards if you could update the member variables to meet the standard we're going for in PhysicsObject and PhysicsSystem that would be much appreciated.

File: CSC8503CoreClasses/PhysicsSystem.cpp [hunk 1]
```diff
@@ -220,51 +221,138 @@ In tutorial 5, we start determining the correct response to a collision,
...
+void PhysicsSystem::FrictionImpulse(GameObject& a, GameObject& b, CollisionDetection::ContactPoint& p, float j) const {
+	PhysicsObject* physA = a.GetPhysicsObject();
+	PhysicsObject* physB = b.GetPhysicsObject();
+	float totalMass = physA->GetInverseMass() + physB->GetInverseMass();
+	Vector3 relativeA = p.localA;
+	Vector3 relativeB = p.localB;
+
+	Vector3 contactVelocity;
+
+	contactVelocity = (GetIsCapsule(a) || GetIsCapsule(b) ? contactVelocity = physB->GetLinearVelocity() - physA->GetLinearVelocity() : contactVelocity = CalculateCollisionVelocity(relativeA, relativeB, physA, physB));
```
Comment (benblack90) [inline CSC8503CoreClasses/PhysicsSystem.cpp:268]: This took a little bit for me to parse: I think you don't need the two 'contactVelocity = ' re-statements in a ternary. So it should read:
contactVelocity = (GetIsCapsule(a) || GetIsCapsule(b)) ? physB->GetLinearVelocity() - physA->GetLinearVelocity : 
                              CalculateCollisionVelocity(relativeA, relativeB, physA, physB);

Which may well still be a little confusing! If-else might be clearer here, but that's up to you.

File: CSC8503CoreClasses/PhysicsSystem.cpp [hunk 2]
```diff
@@ -220,51 +221,138 @@ In tutorial 5, we start determining the correct response to a collision,
...
+	Transform& transformB = b.GetTransform();
+
+	transformA.SetPosition(transformA.GetPosition() - (p.normal * p.penetration * (a.GetPhysicsObject()->GetInverseMass() / totalMass)));
+	transformB.SetPosition(transformB.GetPosition() + (p.normal * p.penetration * (b.GetPhysicsObject()->GetInverseMass() / totalMass)));
+}
+
+bool PhysicsSystem::CheckFrictionShuldBeApplied(float aVelocity, float bVelocity, float totalMass) const {
+	if (aVelocity < 5.5 && bVelocity < 5.5)
+		return false;
+	if (totalMass == 0)
```
Comment (benblack90) [inline CSC8503CoreClasses/PhysicsSystem.cpp:305]: Is there any way mass could (presumably incorrectly) be less than 0? Do we need to check for it?

File: CSC8503CoreClasses/PhysicsSystem.cpp [hunk 3]
```diff
@@ -220,51 +221,138 @@ In tutorial 5, we start determining the correct response to a collision,
...
+	if (obj.GetBoundingVolume()->type == VolumeType::Capsule)
+		return true;
+	return false;
+}
+
+void PhysicsSystem::SeperateObjects(GameObject& a, GameObject& b, CollisionDetection::ContactPoint& p, float totalMass) const {
+	Transform& transformA = a.GetTransform();
+	Transform& transformB = b.GetTransform();
+
+	transformA.SetPosition(transformA.GetPosition() - (p.normal * p.penetration * (a.GetPhysicsObject()->GetInverseMass() / totalMass)));
```
Comment (benblack90) [inline CSC8503CoreClasses/PhysicsSystem.cpp:298]: CheckFrictionShouldBeApplied checks for 0, implying 0 is a possible totalMass. If that's the case, we could be dividing by zero!
If 0 isn't a possibility, maybe we should assert that. If it is, we need to protect this function from the possibility.

**Review by benblack90 — APPROVED** [review]
Just some little nit-picky comments! Mostly fine and swell

**Review by Nijelous — APPROVED** [review]

**Pull request merged**